arm-targets/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+        let string_versions: Vec<String> = [Isa::A64, Isa::A32, Isa::T32]
+            .iter()
+            .map(|i| format!(r#""{i}""#))
+            .collect();
+        string_versions.join(", ")
+    }
+}
+
+impl core::fmt::Display for Isa {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Isa::A64 => "a64",
+                Isa::A32 => "a32",
+                Isa::T32 => "t32",
+            }
+        )
+    }
+}
+
+/// The Arm Architecture
+///
+/// As defined by a particular revision of the Arm Architecture Reference Manual (ARM).
+pub enum Arch {
+    /// Armv6-M (also known as ARMv6-M)
+    Armv6M,
+    /// Armv7-M (also known as ARMv7-M)
+    Armv7M,
+    /// Armv7E-M (also known as ARMv7E-M)
+    Armv7EM,
+    /// Armv8-M Baseline
+    Armv8MBase,
+    /// Armv8-M with Mainline extensions
+    Armv8MMain,
+    /// Armv7-R (also known as ARMv7-R)
+    Armv7R,
+    /// Armv8-R
+    Armv8R,
+    /// Armv7-A (also known as ARMv7-A)
+    Armv7A,
+    /// Armv8-A
+    Armv8A,
+}
+
+impl Arch {
+    /// Decode a target string
+    pub fn get(target: &str) -> Option<Arch> {
+        if target.starts_with("thumbv6m-") {
+            Some(Arch::Armv6M)
+        } else if target.starts_with("thumbv7m-") {
+            Some(Arch::Armv7M)
+        } else if target.starts_with("thumbv7em-") {
+            Some(Arch::Armv7EM)
+        } else if target.starts_with("thumbv8m.base-") {
+            Some(Arch::Armv8MBase)
+        } else if target.starts_with("thumbv8m.main-") {
+            Some(Arch::Armv8MMain)
+        } else if target.starts_with("armv7r-") || target.starts_with("armebv7r") {
+            Some(Arch::Armv7R)
+        } else if target.starts_with("armv8r-") {
+            Some(Arch::Armv8R)
+        } else if target.starts_with("aarch64-") || target.starts_with("aarch64be-") {
+            Some(Arch::Armv8A)
+        } else {
+            None
+        }
+    }
+
+    /// Get the Arm Architecture Profile
+    pub fn profile(&self) -> Profile {
+        match self {
+            Arch::Armv6M | Arch::Armv7M | Arch::Armv7EM | Arch::Armv8MBase | Arch::Armv8MMain => {
+                Profile::M
+            }
+            Arch::Armv7R | Arch::Armv8R => Profile::R,
+            Arch::Armv7A | Arch::Armv8A => Profile::A,
+        }
+    }
+
+    /// Get a comma-separated list of values, suitable for cfg-check
+    pub fn values() -> String {
+        let string_versions: Vec<String> = [
+            Arch::Armv6M,
+            Arch::Armv7M,
+            Arch::Armv7EM,
+            Arch::Armv8MBase,
+            Arch::Armv8MMain,
+            Arch::Armv7R,
+            Arch::Armv8R,
+            Arch::Armv7A,
+            Arch::Armv8A,
+        ]
+        .iter()
+        .map(|i| format!(r#""{i}""#))
+        .collect();
+        string_versions.join(", ")
+    }
+}
+
+impl core::fmt::Display for Arch {
```

Comment:
I find this display implementation quite surprising. There is an `Arch::get(target)` which takes a different format than this display implementation produces.

I guess I'd kind of expect a Display impl to output something similar to what I passed here? As it stands I'd pass eg `thumbv8m.main-***` and get out `v8-m.main` which which alters the format of the 'middle' (from `v8m.main-` to `v8-m.main`) and I'm not really sure why. (I assume there is some specific ARM related reason, which, I get, maybe we can document that here for people like me)

Reply:
The input format is constrained by whatever strings the Rust project cooked up - and many do not follow the official Arm name of the architecture, and conflate architecture and instruction set.

The stringified form of `enum Arch` was intended to reflect what Arm call the architecture, but in lower case, and without spaces. Like, it's actually called "Armv8-M Mainline" and Rust has to drop the '-' because that's the field delimiter.

Reply:
But this hill will not be where I die today, so I'm happy to take proposals for renaming these stringified forms.

Reply:
Okay, that makes a lot of sense. I'm happy with it as-is (A comment may be nice for future readers though). :)